vector.hpp

```diff
@@ -0,0 +1,276 @@
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>::Vector(const T vector[])
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = vector[i];
+}
+
+template <typename T, int dimension>
+template <typename U>
+Vector<T,dimension>::Vector(const Vector<U,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = T(rhs[i]);
+}
+
+template <typename T, int dimension>
+template <typename U>
+Vector<T,dimension>& Vector<T,dimension>::operator=
+(const Vector<U,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = T(rhs[i]);
+  return *this;
+}
+
+template <typename T, int dimension>
+T& Vector<T,dimension>::operator[](int index)
+{
+  return this->elements_[index];
+}
+
+template <typename T, int dimension>
+T Vector<T,dimension>::operator[](int index) const
+{
+  return this->elements_[index];
+}
+
+template <typename T, int dimension>
+Vector<T,dimension> Vector<T,dimension>::operator-() const
+{
+  return (Vector<T,dimension>() - *this);
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator+=(const Vector<T,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] += rhs[i];
+  return *this;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator-=(const Vector<T,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] -= rhs[i];
+  return *this;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator+=(const T rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] += rhs;
+  return *this;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator-=(const T rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] -= rhs;
+  return *this;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator*=(const T rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] *= rhs;
+  return *this;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>& Vector<T,dimension>::operator/=(const T rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] /= rhs;
+  return *this;
+}
+
+template <typename T1, typename T2, int dimension>
+auto operator+(const Vector<T1,dimension>& lhs, const Vector<T2,dimension>& rhs)
+    -> Vector<decltype(lhs[0]+rhs[0]),dimension>
+{
+  Vector<decltype(lhs[0]+rhs[0]),dimension> ans(lhs);
+  ans += rhs;
+  return ans;
+}
+
+template <typename T1, typename T2, int dimension>
+auto operator+(const Vector<T1,dimension>& lhs, const T2 rhs)
+    -> Vector<decltype(lhs[0]+rhs),dimension>
+{
+  Vector<decltype(lhs[0]+rhs),dimension> ans(lhs);
+  ans += rhs;
+  return ans;
+}
+
+template <typename T1, typename T2, int dimension>
+auto operator+(const T1 lhs, const Vector<T2,dimension>& rhs)
+    -> Vector<decltype(lhs+rhs[0]),dimension>
+{
+  Vector<decltype(lhs+rhs[0]),dimension> ans(lhs);
```

Comment:
Does this work? I can't see a constructor that would take a scalar value and initialise all the elements of the vector to that value but you seem to be using such a constructor here.